kotlinx-coroutines-debug/build.gradle.kts

```diff
@@ -0,0 +1,111 @@
+    testImplementation("org.junit.jupiter:junit-jupiter-engine:$junit5_version")
+    testImplementation("org.junit.platform:junit-platform-testkit:1.7.0")
+    add("shadowDeps", "net.bytebuddy:byte-buddy:$byte_buddy_version")
+    add("shadowDeps", "net.bytebuddy:byte-buddy-agent:$byte_buddy_version")
+    compileOnly("io.projectreactor.tools:blockhound:$blockhound_version")
+    testImplementation("io.projectreactor.tools:blockhound:$blockhound_version")
+    testImplementation("com.google.code.gson:gson:2.8.6")
+    api("net.java.dev.jna:jna:$jna_version")
+    api("net.java.dev.jna:jna-platform:$jna_version")
+}
+
+java {
+    /* This is needed to be able to run JUnit5 tests. Otherwise, Gradle complains that it can't find the
+    JVM1.6-compatible version of the `junit-jupiter-api` artifact. */
+    disableAutoTargetJvm()
+}
+
+// This is required for BlockHound tests to work, see https://github.com/Kotlin/kotlinx.coroutines/issues/3701
+tasks.withType<Test>().configureEach {
+    if (JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_13)) {
+        jvmArgs("-XX:+AllowRedefinitionToAddDeleteMethods")
+    }
+}
+
+val jar by tasks.getting(Jar::class) {
+    enabled = false
+}
+
+val shadowJar by tasks.getting(ShadowJar::class) {
+    // Shadow only byte buddy, do not package kotlin stdlib
+    configurations = listOf(project.configurations["shadowDeps"])
+    relocate("net.bytebuddy", "kotlinx.coroutines.repackaged.net.bytebuddy")
+    // exclude the module-info.class provided by bytebuddy
+    exclude("META-INF/versions/9/module-info.class")
+}
+
+// adapted from <https://github.com/johnrengelman/shadow/issues/710#issuecomment-1280585784>
+// There doesn't seem to be a way to ask the shadow jar to contain the `module-info.class` provided by us and not
+// the one provided by bytebuddy. So, this is done in two steps: `shadowJarTask` excludes the module-info from
+// bytebuddy, and this task creates another jar file, this time one that has the correct module-info.
+val shadowJarWithCorrectModuleInfo by tasks.registering(Jar::class) {
+    group = "shadow"
+    from(zipTree(shadowJar.archiveFile))
+    from(tasks.compileModuleInfoJava) {
+        // Include **only** file we are interested in as JavaCompile output also contains some tmp files
+        include("module-info.class")
+        into("META-INF/versions/9/")
+    }
+    duplicatesStrategy = DuplicatesStrategy.FAIL
+    archiveBaseName.set(jar.archiveBaseName)
+    archiveVersion.set(jar.archiveVersion)
+    archiveClassifier.set(jar.archiveClassifier)
+    manifest {
+        attributes(mapOf(
+            "Premain-Class" to "kotlinx.coroutines.debug.AgentPremain",
+            "Can-Redefine-Classes" to "true",
+            "Multi-Release" to "true"
+        ))
+    }
+}
+
+tasks.getByName("publishMavenPublicationToMavenLocal") {
+    dependsOn(shadowJarWithCorrectModuleInfo)
+}
+
+VersionFile.configure(project, shadowJarWithCorrectModuleInfo.get())
+
+configurations.all {
```

Comment:
Is it correct that this block removes everything from the output that depends on the default `ShadowJar` task, and replaces everything that contains the dependency on `Jar` task with `ShadowJarWithCorrectModuleInfo`?

Reply:
I'm a bit lost here as the `removeIf` covers all artifacts that depends on our jars and the replacement can also replace artifacts in more than one task

Reply:
This block has the following effect on the configurations that have any outgoing artifacts to begin with:
```
':kotlinx-coroutines-debug:apiElements' (before): [jar]
':kotlinx-coroutines-debug:apiElements' (after): [endResultJar]
':kotlinx-coroutines-debug:archives' (before): [jar]
':kotlinx-coroutines-debug:archives' (after): [endResultJar]
':kotlinx-coroutines-debug:mainSourceElements' (before): [DecoratingPublishArtifact_Decorated resources:directory::null, DecoratingPublishArtifact_Decorated java:directory::null, DecoratingPublishArtifact_Decorated src:directory::null]
':kotlinx-coroutines-debug:mainSourceElements' (after): [DecoratingPublishArtifact_Decorated resources:directory::null, DecoratingPublishArtifact_Decorated java:directory::null, DecoratingPublishArtifact_Decorated src:directory::null]
':kotlinx-coroutines-debug:runtimeElements' (before): [jar]
':kotlinx-coroutines-debug:runtimeElements' (after): [endResultJar]
':kotlinx-coroutines-debug:shadow' (before): [shadowJar]
':kotlinx-coroutines-debug:shadow' (after): []
':kotlinx-coroutines-debug:shadowRuntimeElements' (before): [shadowJar]
':kotlinx-coroutines-debug:shadowRuntimeElements' (after): []
':kotlinx-coroutines-debug:testResultsElementsForTest' (before): [DecoratingPublishArtifact_Decorated binary:directory::null]
':kotlinx-coroutines-debug:testResultsElementsForTest' (after): [DecoratingPublishArtifact_Decorated binary:directory::null]
```